homeassistant/components/sensor/nextbus.py

```diff
@@ -0,0 +1,271 @@
+
+
+def validate_value(value_name, value, value_list):
+    """Validate tag value is in the list of items and logs error if not."""
+    valid_values = {
+        v['tag']: v['title']
+        for v in value_list
+    }
+    if value not in valid_values:
+        _LOGGER.error(
+            'Invalid %s tag `%s`. Please use one of the following: %s',
+            value_name,
+            value,
+            ', '.join(
+                '{}: {}'.format(title, tag)
+                for tag, title in valid_values.items()
+            )
+        )
+        return False
+
+    return True
+
+
+def validate_tags(agency, route, stop):
+    """Validate provided tags."""
+    from py_nextbus import NextBusClient  # pylint: disable=import-error
+    client = NextBusClient(output_format='json')
+
+    # Validate agencies
+    if not validate_value(
+        'agency',  # pylint: disable=C0330
+        agency,  # pylint: disable=C0330
+        client.get_agency_list()['agency'],  # pylint: disable=C0330
+    ):
+        return False
+
+    # Validate the route
+    if not validate_value(
+        'route',  # pylint: disable=C0330
+        route,  # pylint: disable=C0330
+        client.get_route_list(agency)['route'],  # pylint: disable=C0330
+    ):
+        return False
+
+    # Validate the stop
+    route_config = client.get_route_config(route, agency)['route']
+    if not validate_value(
+        'stop',  # pylint: disable=C0330
+        stop,  # pylint: disable=C0330
+        route_config['stop'],  # pylint: disable=C0330
+    ):
+        return False
+
+    return True
+
+
+def setup_platform(hass, config, add_entities, discovery_info=None):
+    """Load values from configuration and initialize the platform."""
+    agency = config.get(CONF_AGENCY)
+    route = config.get(CONF_ROUTE)
+    stop = config.get(CONF_STOP)
+    name = config.get(CONF_NAME)
+
+    # Ensures that the tags provided are valid, also logs out valid values
+    if not validate_tags(agency, route, stop):
+        raise ValueError('Invalid config value(s)')
+
+    add_entities([
+        NextBusDepartureSensor(
+            agency,
+            route,
+            stop,
+            name,
+        ),
+    ], True)
+
+
+class NextBusDepartureSensor(Entity):
+    """Sensor class that displays upcoming NextBus times.
+
+    To function, this requires knowing the agency tag as well as the tags for
+    both the route and the stop.
+
+    This is possibly a little convoluted to provide as it requires making a
+    request to the service to get these values. Perhaps it can be simplifed in
+    the future using fuzzy logic and matching.
+    """
+
+    def __init__(self, agency, route, stop, name=None):
+        """Initialize sensor with all required config."""
+        self.agency = agency
+        self.route = route
+        self.stop = stop
+        self._custom_name = name
+        self._name = '{} - {}'.format(agency, route)
```

Comment:
Don't use space and dash next to one another in the name. They will be converted to underscore making the entity_id invalid. Just separate words by a single space in the name.

Reply:
Is there a way to specify the `entity_id` and `name` separately? Removing the hyphen will sacrifice readability.

Reply:
We prefer not to do that.